internal/pkg/source/source.go

```diff
@@ -130,10 +132,38 @@ func (a *Source) visitReferrers(referrers *[]ssa.Instruction) {
 				continue
 			}
 		}
+
+		// If the referrer is in a different block from the one we last visited,
+		// and it can't be reached from the block we are visiting, then stop visiting.
+		rb := r.Block()
+		if a.lastBlockVisited != nil && rb != a.lastBlockVisited && !a.canReach(a.lastBlockVisited, rb) {
+			continue
+		}
+
 		a.dfs(r.(ssa.Node))
 	}
 }
 
+func (a *Source) canReach(start *ssa.BasicBlock, dest *ssa.BasicBlock) bool {
+	stack := []*ssa.BasicBlock{start}
+	seen := map[*ssa.BasicBlock]bool{start: true}
+	for len(stack) > 0 {
+		current := stack[len(stack)-1]
```

Comment:
That's true, but it is defined like so: `type stack []ssa.Node`, so it can't be used here. I could introduce a new type that would be defined like `type stack []*ssa.BasicBlock`. The definitions for its methods would be identical to those of the existing `stack` type, though.

Alternatively, I could write a generic `stack`, but I feel like that wouldn't be a good ROI. We've had this discussion in the past, but I don't think having a dedicated `stack` type is a big win.

Reply:
Up to you.
IMHO, implementing generic algorithms (inline) within a function that is clearly domain specific (SSA) increases the cognitive overhead for review. So, I would try to hide this.

Reply:
SG, I've duplicated the `stack` code and modified `canReach` to use it.